File: dashboards/dashboards/page/sales_dashboard/sales_dashboard.py
from __future__ import annotations

from typing import Any

import frappe
from frappe.utils import flt, getdate, today

from dashboards.dashboards.dashboard_data import MONTH_LABELS, get_item_cogs_map, get_item_rcp_map


MONTHS = [{"key": label.lower(), "label": label} for label in MONTH_LABELS]
MONTH_MAP = {item["key"]: index + 1 for index, item in enumerate(MONTHS)}


def _get_years() -> list[str]:
	rows = frappe.db.sql(
		"""
		SELECT DISTINCT YEAR(posting_date) AS year
		FROM `tabSales Invoice`
		WHERE docstatus = 1
		  AND COALESCE(is_return, 0) = 0
		  AND posting_date IS NOT NULL
		ORDER BY year
		""",
		as_dict=True,
	)

	years = [str(row.year) for row in rows if row.year]
	if years:
		return years

	return [str(getdate(today()).year)]


def _get_default_period() -> tuple[str, str]:
	latest_row = frappe.db.sql(
		"""
		SELECT MAX(posting_date) AS posting_date
		FROM `tabSales Invoice`
		WHERE docstatus = 1
		  AND COALESCE(is_return, 0) = 0
		""",
		as_dict=True,
	)[0]

	reference_date = getdate(latest_row.posting_date) if latest_row.posting_date else getdate(today())
	return str(reference_date.year), MONTH_LABELS[reference_date.month - 1].lower()


def _normalize_filters(year: str | None, month: str | None) -> tuple[str, str]:
	years = _get_years()
	default_year, default_month = _get_default_period()
	selected_year = year if year in years else default_year
	selected_month = month if month in MONTH_MAP else default_month
	return selected_year, selected_month


def _get_product_rows(year: str, month: str) -> list[dict[str, Any]]:
	cogs_map = get_item_cogs_map(year, MONTH_MAP[month])
	rcp_map = get_item_rcp_map(year, MONTH_MAP[month])
	rows = frappe.db.sql(
		"""
		SELECT
			COALESCE(NULLIF(sii.item_code, ''), NULLIF(sii.item_name, ''), 'Unknown Item') AS item_key,
			COALESCE(NULLIF(sii.item_name, ''), sii.item_code, 'Unknown Item') AS item,
			SUM(COALESCE(sii.stock_qty, sii.qty, 0)) AS kg,
			SUM(COALESCE(sii.base_net_amount, sii.net_amount, sii.base_amount, sii.amount, 0)) AS sales,
			SUM(COALESCE(sii.stock_qty, sii.qty, 0) * COALESCE(sii.incoming_rate, 0)) AS cost,
			0 AS rsp
		FROM `tabSales Invoice` si
		INNER JOIN `tabSales Invoice Item` sii ON sii.parent = si.name
		WHERE si.docstatus = 1
		  AND COALESCE(si.is_return, 0) = 0
		  AND YEAR(si.posting_date) = %(year)s
		  AND MONTH(si.posting_date) = %(month)s
		GROUP BY
			COALESCE(NULLIF(sii.item_code, ''), NULLIF(sii.item_name, ''), 'Unknown Item'),
			COALESCE(NULLIF(sii.item_name, ''), sii.item_code, 'Unknown Item')
		ORDER BY sales DESC, item ASC
		""",
		{"year": int(year), "month": int(MONTH_MAP[month])},
		as_dict=True,
	)

	result = []
	for row in rows:
		sales = flt(row.sales)
		cost = flt(row.cost) or flt(cogs_map.get(row.item_key))
		margin = sales - cost
		rsp = flt(rcp_map.get(row.item_key))
		profit = margin - rsp
		result.append(
			{
				"item": row.item,
				"kg": round(flt(row.kg)),
				"sales": round(sales),
				"cost": round(cost),
				"margin": round(margin),
				"rsp": round(rsp),
				"margin_percent": (margin / sales * 100) if sales else 0,
				"profit": round(profit),
			}
		)

	return result


@frappe.whitelist()
def get_dashboard_context(year: str | None = None, month: str | None = None):
	selected_year, selected_month = _normalize_filters(year, month)

	return {
		"default_filters": {
			"year": selected_year,
			"month": selected_month,
		},
		"years": _get_years(),
		"months": MONTHS,
		"product_rows": _get_product_rows(selected_year, selected_month),
	}

File: dashboards/dashboards/page/sales_dashboard/sales_dashboard.js
frappe.pages["sales-dashboard"].on_page_load = function (wrapper) {
	new dashboards.ui.SalesDashboardPage(wrapper);
};

frappe.provide("dashboards.ui");

dashboards.ui.SalesDashboardPage = class SalesDashboardPage {
	constructor(wrapper) {
		this.wrapper = $(wrapper);
		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: __("Панель продаж"),
			single_column: true,
		});

		this.state = {
			year: null,
			month: null,
		};

		this.make_layout();
		this.load_context();
	}

	ensure_runtime_styles() {
		if (document.getElementById("sales-dashboard-table-layout-fix")) {
			return;
		}

		const style = document.createElement("style");
		style.id = "sales-dashboard-table-layout-fix";
		style.textContent = `
			.sales-dashboard-table-panel {
				width: min(100%, 980px) !important;
				max-width: calc(100% - 20px) !important;
				margin: 0 10px 0 auto !important;
				transform: none !important;
				box-sizing: border-box !important;
			}
			.sales-dashboard-table-wrap {
				max-width: 100% !important;
				overflow: auto !important;
			}
			.sales-dashboard-table {
				width: max-content !important;
				min-width: 980px !important;
				table-layout: auto !important;
				border-collapse: collapse !important;
			}
			.sales-dashboard-table .is-text {
				min-width: 280px !important;
				text-align: left !important;
			}
			.sales-dashboard-table .is-number {
				min-width: 112px !important;
				text-align: right !important;
				white-space: nowrap !important;
			}
		`;
		document.head.appendChild(style);
	}

	make_layout() {
		this.ensure_runtime_styles();
		this.wrapper.find(".layout-main-section-wrapper").addClass("sales-dashboard-layout");
		this.wrapper.find(".page-head").addClass("sales-dashboard-page-head");
		this.page.main.removeClass("frappe-card");

		this.page.main.html(`
			<div class="sales-dashboard-screen">
				<div class="sales-dashboard-shell">
					<div class="sales-dashboard-filters sales-dashboard-filters--years" data-region="years"></div>
					<div class="sales-dashboard-filters sales-dashboard-filters--months" data-region="months"></div>
					<div class="sales-dashboard-table-panel">
						<div class="sales-dashboard-table-wrap" data-region="table"></div>
					</div>
				</div>
			</div>
		`);

		dashboards.ui.setupDashboardSidebar({
			page: this.page,
			route: "sales-dashboard",
		});

		this.$years = this.page.main.find('[data-region="years"]');
		this.$months = this.page.main.find('[data-region="months"]');
		this.$table = this.page.main.find('[data-region="table"]');
	}

	load_context() {
		this.render_loading();

		frappe.call({
			method: "dashboards.dashboards.page.sales_dashboard.sales_dashboard.get_dashboard_context",
			args: {
				year: this.state.year,
				month: this.state.month,
			},
			callback: (r) => {
				this.context = r.message || {};
				this.state = { ...(this.context.default_filters || {}) };
				this.render();
			},
		});
	}

	render() {
		this.render_years();
		this.render_months();
		this.render_table();
	}

	render_years() {
		const years = this.context.years || [];
		const hasOverflow = years.length > 6;
		const wrapClass = hasOverflow ? "is-scrollable" : "is-centered";
		this.$years.html(
			`
				<div class="sales-dashboard-year-group-wrap ${wrapClass}">
					<div class="sales-dashboard-year-group">
						${years
							.map(
								(year) => `
									<button class="sales-dashboard-chip sales-dashboard-chip--year ${year === this.state.year ? "is-active" : ""}" data-year="${year}">
										${frappe.utils.escape_html(year)}
									</button>
								`
							)
							.join("")}
					</div>
				</div>
			`
		);

		this.$years.find("[data-year]").on("click", (e) => {
			this.state.year = String($(e.currentTarget).data("year"));
			this.load_context();
		});
	}

	render_months() {
		const months = this.context.months || [];
		this.$months.html(
			months
				.map(
					(month) => `
						<button class="sales-dashboard-chip sales-dashboard-chip--month ${month.key === this.state.month ? "is-active" : ""}" data-month="${month.key}">
							${frappe.utils.escape_html(month.label)}
						</button>
					`
				)
				.join("")
		);

		this.$months.find("[data-month]").on("click", (e) => {
			this.state.month = String($(e.currentTarget).data("month"));
			this.load_context();
		});
	}

	render_table() {
		const rows = this.context.product_rows || [];
		const total = this.buildTotalRow(rows);
		const headers = ["Товары", "КГ", "Сумма продаж", "Себестоимость", "Маржа", "Сумма РСП", "Рен", "Прибыль"];

		if (!rows.length) {
			this.$table.html(`
				<div class="sales-dashboard-table-empty">
					${__("За выбранный период нет проведенных счетов продаж.")}
				</div>
			`);
			return;
		}

		this.$table.html(`
			<table class="sales-dashboard-table">
				<thead>
					<tr>${headers.map((header, index) => `<th class="${index === 0 ? "is-text" : "is-number"}">${header}</th>`).join("")}</tr>
				</thead>
				<tbody>
					${rows
						.map(
							(row) => `
								<tr>
									<td class="is-text">${frappe.utils.escape_html(row.item)}</td>
									<td class="is-number">${this.formatInteger(row.kg)}</td>
									<td class="is-number">${this.formatInteger(row.sales)}</td>
									<td class="is-number">${this.formatInteger(row.cost)}</td>
									<td class="is-number">${this.formatInteger(row.margin)}</td>
									<td class="is-number">${this.formatInteger(row.rsp)}</td>
									<td class="is-number">${this.formatPercent(row.margin_percent)}</td>
									<td class="is-number ${row.profit < 0 ? "is-negative" : ""}">${this.formatInteger(row.profit)}</td>
								</tr>
							`
						)
						.join("")}
					<tr class="is-total">
						<td class="is-text">Итого</td>
						<td class="is-number">${this.formatInteger(total.kg)}</td>
						<td class="is-number">${this.formatInteger(total.sales)}</td>
						<td class="is-number">${this.formatInteger(total.cost)}</td>
						<td class="is-number">${this.formatInteger(total.margin)}</td>
						<td class="is-number">${this.formatInteger(total.rsp)}</td>
						<td class="is-number">${this.formatPercent(total.margin_percent)}</td>
						<td class="is-number ${total.profit < 0 ? "is-negative" : ""}">${this.formatInteger(total.profit)}</td>
					</tr>
				</tbody>
			</table>
		`);
	}

	buildTotalRow(rows) {
		const total = rows.reduce(
			(accumulator, row) => {
				accumulator.kg += row.kg;
				accumulator.sales += row.sales;
				accumulator.cost += row.cost;
				accumulator.margin += row.margin;
				accumulator.rsp += row.rsp;
				accumulator.profit += row.profit;
				return accumulator;
			},
			{ kg: 0, sales: 0, cost: 0, margin: 0, rsp: 0, profit: 0 }
		);

		total.margin_percent = total.sales ? (total.margin / total.sales) * 100 : 0;
		return total;
	}

	formatInteger(value) {
		const sign = value < 0 ? "-" : "";
		const numeric = Math.abs(Math.round(value));
		return `${sign}${String(numeric).replace(/\B(?=(\d{3})+(?!\d))/g, " ")}`;
	}

	formatPercent(value) {
		return `${Number(value || 0).toFixed(1).replace(".", ",")}%`;
	}

	render_loading() {
		this.$table.html(`<div class="sales-dashboard-table-empty">${__("Загрузка...")}</div>`);
	}
};
